pallets/corporate-actions/src/distribution.rs

```diff
@@ -0,0 +1,506 @@
+    /// Whether the CAA has claimed remaining funds.
+    pub reclaimed: bool,
+    /// An optional timestamp of payout start
+    pub payment_at: Moment,
+    /// An optional timestamp for payout end
+    pub expires_at: Option<Moment>,
+}
+
+/// Has the distribution expired?
+fn expired(expiry: Option<Moment>, now: Moment) -> bool {
+    expiry.filter(|&e| e <= now).is_some()
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as CapitalDistribution {
+        /// All capital distributions, tied to their respective corporate actions (CAs).
+        ///
+        /// (CAId) => Distribution
+        Distributions get(fn distributions): map hasher(blake2_128_concat) CAId => Option<Distribution<T::Balance>>;
+
+        /// Has an asset holder been paid yet?
+        ///
+        /// (CAId, DID) -> Was DID paid in the CAId?
+        HolderPaid get(fn holder_paid): map hasher(blake2_128_concat) (CAId, IdentityId) => bool;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Start and attach a capital distribution, to the CA identified by `ca_id`,
+        /// with `amount` funds in `currency` withdrawn from `portfolio` belonging to `origin`'s DID.
+        ///
+        /// The distribution will commence at `payment_at` and expire at `expires_at`,
+        /// if provided, or if `None`, then there's no expiry.
+        ///
+        /// ## Arguments
+        /// - `origin` which must be a signer for the CAA of `ca_id`.
+        /// - `ca_id` identifies the CA to start a capital distribution for.
+        /// - `portfolio` specifies the portfolio number of the CAA to distribute `amount` from.
+        /// - `currency` to withdraw and distribute from the `portfolio`.
+        /// - `amount` of `currency` to withdraw and distribute.
+        /// - `payment_at` specifies when benefits may first be pushed or claimed.
+        /// - `expires_at` specifies, if provided, when remaining benefits are forfeit
+        ///    and may be reclaimed by `origin`.
+        ///
+        /// # Errors
+        /// - `UnauthorizedAsAgent` if `origin` is not `ticker`'s sole CAA (owner is not necessarily the CAA).
+        /// - `DistributingAsset` if `ca_id.ticker == currency`.
+        /// - `ExpiryBeforePayment` if `expires_at.unwrap() <= payment_at`.
+        /// - `NowAfterPayment` if `payment_at < now`.
+        /// - `NoSuchCA` if `ca_id` does not identify an existing CA.
+        /// - `NoRecordDate` if CA has no record date.
+        /// - `RecordDateAfterStart` if CA's record date > payment_at.
+        /// - `InsufficientPortfolioBalance` if `portfolio` has less than `amount` of `currency`.
+        /// - `InsufficientBalance` if the protocol fee couldn't be charged.
+        #[weight = 2_000_000_000]
+        pub fn distribute(
+            origin,
+            ca_id: CAId,
+            portfolio: PortfolioNumber,
```

Comment:
This should be optional. Default portfolios do not have a portfolio number attached to them.